### PR TITLE
Fix invalid wasi targets compatibility

### PR DESCRIPTION
**Pull request by antaalt**
Thanks for your help following [Fix WASI compilation for C++](https://github.com/rust-lang/cc-rs/pull/1083), I have tested it on my side from main and spotted some issues with incompatible targets, took some time to clean this aswell and ensure every targets are clearly defined to avoid potential issues with no-exceptions flags

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -3922,6 +3921,16 @@ impl Build {
...
     }
+    fn is_wasi_target(target: &str) -> bool {
+        const TARGETS: [&'static str; 5] = [
+            "wasm32-wasi",
+            "wasm32-wasip1",
+            "wasm32-wasip1-threads",
+            "wasm32-wasip2",
+            "wasm32-wasi-threads",
+        ];
+        return TARGETS.contains(&target);
```
Comment (madsmtm) [inline src/lib.rs:3932]: Why are we listing each target? Would `target.contains("wasi")` not be more forwards-compatible?

Comment (antaalt) [reply src/lib.rs:3932]: Maybe, it was mostly to ensure if they had a target supporting exceptions or something like that to handle them separately, but maybe its overkill

File: src/lib.rs [hunk 2]
```diff
@@ -1334,11 +1335,12 @@ impl Build {
                     .print_metadata(&format_args!("cargo:rustc-link-lib={}", stdlib.display()));
             }
             // Link c++ lib from WASI sysroot
-            if self.get_target()?.contains("wasi") {
+            if Build::is_wasi_target(target.as_ref()) {
                 let wasi_sysroot = self.wasi_sysroot()?;
                 self.cargo_output.print_metadata(&format_args!(
-                    "cargo:rustc-flags=-L {}/lib/wasm32-wasi -lstatic=c++ -lstatic=c++abi",
-                    Path::new(&wasi_sysroot).display()
+                    "cargo:rustc-flags=-L {}/lib/{} -lstatic=c++ -lstatic=c++abi",
```
Comment (NobodyXu) [inline src/lib.rs:1341]: Can you explain a bit about this change please?

In your previous PR it was hard coded, and now it changes to wasm32-wasi

**Review by NobodyXu — COMMENTED** [review]
Thank you!

BTW, cc @antaalt  it'd be great if we can have some testing for wasi targets!

File: src/lib.rs (hunk 2 above)
Comment (NobodyXu) [reply src/lib.rs:1341]: Hmmm maybe I'm just thinking too much, it's probably because we expanded the targets supported.

Comment (antaalt) [reply src/lib.rs:1341]: Yes, the lib folder structure in sysroot looks like this: image

**Comment by antaalt** [comment]
> Thank you!
> 
> BTW, cc @antaalt it'd be great if we can have some testing for wasi targets!

Yes, I will look into it

**Pull request merged**